File: livingrimoire start here/LivinGrimoire java/src/Auxiliary_Modules/ElizaDeducer.java
package Auxiliary_Modules;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ElizaDeducer {
    /*
    * this class populates a special chat dictionary
    * based on the matches added via its addPhraseMatcher function
    * see subclass ElizaDeducerInitializer for example:
    * ElizaDeducer ed = new ElizaDeducerInitializer(2); // 2 = limit of replies per input
    * */
    public List<PhraseMatcher> babble2;
    private final java.util.Map<String, List<PhraseMatcher>> patternIndex;
    private final java.util.Map<String, List<AXKeyValuePair>> responseCache;
    private final EventChatV2 ec2; // chat dictionary, use getter for access. hardcoded replies can also be added

    public ElizaDeducer(int lim) {
        babble2 = new ArrayList<>();
        patternIndex = new java.util.HashMap<>();
        responseCache = new java.util.HashMap<>();
        ec2 = new EventChatV2(lim);
    }

    public EventChatV2 getEc2() {
        return ec2;
    }

    public void learn(String msg) {
        // populate EventChat dictionary
        // Check cache first
        if (responseCache.containsKey(msg)) {
            ec2.addKeyValues(new ArrayList<>(responseCache.get(msg)));
        }

        // Search for matching patterns
        List<PhraseMatcher> potentialMatchers = getPotentialMatchers(msg);
        for (PhraseMatcher pm : potentialMatchers) {
            if (pm.matches(msg)) {
                ArrayList<AXKeyValuePair> response = pm.respond(msg);
                responseCache.put(msg, response);
                ec2.addKeyValues(response);
            }
        }
    }
    public boolean learnedBool(String msg) {
        // same as learn method but returns true if it learned new replies
        boolean learned = false;
        // populate EventChat dictionary
        // Check cache first
        if (responseCache.containsKey(msg)) {
            ec2.addKeyValues(new ArrayList<>(responseCache.get(msg)));
            learned = true;
        }

        // Search for matching patterns
        List<PhraseMatcher> potentialMatchers = getPotentialMatchers(msg);
        for (PhraseMatcher pm : potentialMatchers) {
            if (pm.matches(msg)) {
                ArrayList<AXKeyValuePair> response = pm.respond(msg);
                responseCache.put(msg, response);
                ec2.addKeyValues(response);
                learned = true;
            }
        }
        return learned;
    }
    public String respond(String str1){
        return ec2.response(str1);
    }
    public String respondLatest(String str1){
        // get most recent reply/data
        return ec2.responseLatest(str1);
    }

    private List<PhraseMatcher> getPotentialMatchers(String msg) {
        List<PhraseMatcher> potentialMatchers = new ArrayList<>();
        for (String key : patternIndex.keySet()) {
            if (msg.contains(key)) {
                potentialMatchers.addAll(patternIndex.get(key));
            }
        }
        return potentialMatchers;
    }

    public void addPhraseMatcher(String pattern, String... kvPairs) {
        ArrayList<AXKeyValuePair> kvs = new ArrayList<>();
        for (int i = 0; i < kvPairs.length; i += 2) {
            kvs.add(new AXKeyValuePair(kvPairs[i], kvPairs[i + 1]));
        }
        PhraseMatcher matcher = new PhraseMatcher(pattern, kvs);
        babble2.add(matcher);
        indexPattern(pattern, matcher);
    }

    private void indexPattern(String pattern, PhraseMatcher matcher) {
        for (String word : pattern.split("\\s+")) {
            patternIndex.computeIfAbsent(word, k -> new ArrayList<>()).add(matcher);
        }
    }

    public static class PhraseMatcher {
        public final Pattern matcher;
        public final List<AXKeyValuePair> responses;

        public PhraseMatcher(String matcher, List<AXKeyValuePair> responses) {
            this.matcher = Pattern.compile(matcher);
            this.responses = responses;
        }

        public boolean matches(String str) {
            Matcher m = matcher.matcher(str);
            return m.matches();
        }

        public ArrayList<AXKeyValuePair> respond(String str) {
            Matcher m = matcher.matcher(str);
            ArrayList<AXKeyValuePair> result = new ArrayList<>();
            if (m.find()) {
                int tmp = m.groupCount();
                for (AXKeyValuePair kv : this.responses) {
                    AXKeyValuePair tempKV = new AXKeyValuePair(kv.getKey(), kv.getValue());
                    for (int i = 0; i < tmp; i++) {
                        String s = m.group(i + 1);
                        tempKV.setKey(tempKV.getKey().replace("{" + i + "}", s).toLowerCase());
                        tempKV.setValue(tempKV.getValue().replace("{" + i + "}", s).toLowerCase());
                    }
                    result.add(tempKV);
                }
            }
            return result;
        }
    }
}

File: livingrimoire start here/LivinGrimoire Kotlin/kotlin/auxiliary_modules/ElizaDeducer.kt
package auxiliary_modules

import java.util.*
import java.util.regex.Pattern
import kotlin.collections.ArrayList

open class ElizaDeducer(lim: Int) {
    /*
    * this class populates a special chat dictionary
    * based on the matches added via its addPhraseMatcher function
    * see subclass ElizaDeducerInitializer for example:
    * ElizaDeducer ed = new ElizaDeducerInitializer(2); // 2 = limit of replies per input
    * */
    var babble2: MutableList<PhraseMatcher>
    private val patternIndex: MutableMap<String, MutableList<PhraseMatcher>>
    private val responseCache: MutableMap<String, List<AXKeyValuePair>>
    val ec2 // chat dictionary, use getter for access. hardcoded replies can also be added
            : EventChatV2

    init {
        babble2 = ArrayList()
        patternIndex = HashMap()
        responseCache = HashMap()
        ec2 = EventChatV2(lim)
    }

    fun learn(msg: String) {
        // populate EventChat dictionary
        // Check cache first
        if (responseCache.containsKey(msg)) {
            ec2.addKeyValues(ArrayList(responseCache[msg]!!))
        }

        // Search for matching patterns
        val potentialMatchers = getPotentialMatchers(msg)
        for (pm in potentialMatchers) {
            if (pm.matches(msg)) {
                val response = pm.respond(msg)
                responseCache[msg] = response
                ec2.addKeyValues(response)
            }
        }
    }

    fun learnedBool(msg: String): Boolean {
        // same as learn method but returns true if it learned new replies
        var learned = false
        // populate EventChat dictionary
        // Check cache first
        if (responseCache.containsKey(msg)) {
            ec2.addKeyValues(ArrayList(responseCache[msg]!!))
            learned = true
        }

        // Search for matching patterns
        val potentialMatchers = getPotentialMatchers(msg)
        for (pm in potentialMatchers) {
            if (pm.matches(msg)) {
                val response = pm.respond(msg)
                responseCache[msg] = response
                ec2.addKeyValues(response)
                learned = true
            }
        }
        return learned
    }

    fun respond(str1: String?): String {
        return ec2.response(str1!!)
    }

    fun respondLatest(str1: String?): String {
        // get most recent reply/data
        return ec2.responseLatest(str1!!)
    }

    private fun getPotentialMatchers(msg: String): List<PhraseMatcher> {
        val potentialMatchers: MutableList<PhraseMatcher> = ArrayList()
        for (key in patternIndex.keys) {
            if (msg.contains(key)) {
                potentialMatchers.addAll(patternIndex[key]!!)
            }
        }
        return potentialMatchers
    }

    fun addPhraseMatcher(pattern: String, vararg kvPairs: String?) {
        val kvs = ArrayList<AXKeyValuePair>()
        var i = 0
        while (i < kvPairs.size) {
            kvs.add(AXKeyValuePair(kvPairs[i]!!, kvPairs[i + 1]!!))
            i += 2
        }
        val matcher = PhraseMatcher(pattern, kvs)
        babble2.add(matcher)
        indexPattern(pattern, matcher)
    }

    private fun indexPattern(pattern: String, matcher: PhraseMatcher) {
        for (word in pattern.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            patternIndex.computeIfAbsent(
                word
            ) { k: String? -> ArrayList() }.add(matcher)
        }
    }

    class PhraseMatcher(matcher: String, val responses: List<AXKeyValuePair>) {
        val matcher: Pattern

        init {
            this.matcher = Pattern.compile(matcher)
        }

        fun matches(str: String?): Boolean {
            val m = matcher.matcher(str)
            return m.matches()
        }

        fun respond(str: String): ArrayList<AXKeyValuePair> {
            val m = matcher.matcher(str)
            val result = ArrayList<AXKeyValuePair>()
            if (m.find()) {
                val tmp = m.groupCount()
                for (kv in responses) {
                    val tempKV = AXKeyValuePair(kv.key, kv.value)
                    for (i in 0 until tmp) {
                        val s = m.group(i + 1)
                        tempKV.key = tempKV.key.replace("{$i}", s).lowercase(Locale.getDefault())
                        tempKV.value = tempKV.value.replace("{$i}", s).lowercase(Locale.getDefault())
                    }
                    result.add(tempKV)
                }
            }
            return result
        }
    }
}